### ISSUE TITLE
ignore parameter assignments to specific variables with no-param-reassign enabled

### ISSUE THREAD

**Issue by TobiasBales** [eslint/eslint#6505]
**What version of ESLint are you using?**
2.11.1

With angular you often see patterns like 

``` js
app.controller('MyCtrl', function($scope) {
  $scope.something = true;
});
```

It would be nice to be able to setup eslint to strictly lint everything but ignore assignments to $scope (if configured to ignore $scope)

**Comment by eslintbot** [comment]
Hi @TobiasBales, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by mysticatea** [comment]
Thank you for this issue.

Though you didn't write your configuration, I'm guessing you are using `no-param-reassign` with `{props: true}` option.
Do you think that disabling `{props: true}` is not enough?

**Comment by TobiasBales** [comment]
I think it is the wrong idea to generally make the rule less strict (with configuring it that way).

I do think, that reassigning props to function parameters should be avoided or just done if it can't be helped (eg angulars $scope) but I would love to be able to enumerate the exceptions I want and still get errors when I assign to props of other variables.

**Comment by aubergine10** [comment]
I've got a very similar scenario in one of my apps - for performance reasons a subset of functions pass and update objects amongst themselves. I don't want to turn the rule off because it should apply everywhere else. I don't want to clutter the code with eslint block comments. Being able to define specific exclusions would be preferable, especially because I could put comments in the `.eslintrc` explaining _why_ that specific param is excluded from the rule.

**Comment by TobiasBales** [comment]
Any update or thoughts on this?

**Comment by ilyavolodin** [comment]
@TobiasBales I think in your case, it's best to setup this rule as `no-param-reassign: ["error", { "props": false }]`, your case is very specific to Angular, and I don't think it's generic enough to worriment adding exception into the core. If anything, `eslint-plugin-angular` should clone `no-param-reassign` and modify it to ignore `$scope`.

Overall I'm :-1: for adding this into the core.

**Comment by platinumazure** [comment]
I think this is something the team can consider, but we will need a concrete proposal for how it would be done.

Specifically, we would need to know what new configurations we would support for the rule, along with a clear explanation of what new invalid code or valid code would arise (compared to the rule's current behavior).

Once we have that proposal, we can follow our [rule change process](http://eslint.org/docs/developer-guide/contributing/rule-changes) by looking for a champion and consensus among the team, which will then allow us to accept this proposed enhancement. At that point, our champion will work with the issue submitter (or anyone else interested in making the change) in order to implement the change.

**Comment by TobiasBales** [comment]
My example was angular specific, but as @aubergine10 mentioned express apps are the same, same goes for some chrome apis (file write access for extensions for example)

As for what code is the valid that wasn't before etc, I opened a pull request that does implement that feature, basics can be seen from the test cases I wrote but I will gladly formalize it as a proposal.
I would gladly write a proposal, is there an outline on how it would look like?

**Comment by TobiasBales** [comment]
My proposed changes are:
add an optional array with to the configuration of `no-param-reassigned` called `ignorePropertiesFor`
It contains strings naming variables for which property assignments/modifications will be ignored.

The configuration could look like:

```json
{
  "rules": {
    "no-param-reassign": ["error", { "ignorePropertiesFor": ["angular", "res"] } ]
  }
}
```

With that configuration the following code would not produce errors:
```js
app.controller('MyCtrl', function($scope) {
  $scope.something = true;
});

function someController(req, res, next) {
    res.statusCode = 404;
}
```

but this would report the usual property assignment errors:
```js
function someFunction(randomParameter) {
  randomParameter.valid = false;
}
```

If one would just use `["error", { "proprs": false }]` as configuration the last example and similar things would not be caught which is why I would like to propose that change and also implemented it already.

**Comment by aubergine10** [comment]
Random thought: The view seems to be that having exceptions in rules is a bad thing, however I feel the opposite is true. Someone who doesn't want the rule in the first place will disable the rule. People who do want the rule, but are stymied by certain platform features have more of a struggle to implement the rule. By adding in some flexibility, the rule can be adopted where possible, rather than having to be completely disabled. Best practice is the ideal, but where that can't be achieved it's better to have "Better practice where possible" than "Worst practice everywhere". We might not be able to work around platform specific patterns, but we can at least prevent those patterns creeping in to the rest of our code.

**Comment by nzakas** [comment]
@eslint/eslint-team looks like we lost track of this. Does anyone want to champion this proposal?

**Comment by TobiasBales** [comment]
@nzakas what exactly does champion in this case mean?
Take responsibility? That would be great.
Implement it? I already did that including tests - see #6504

**Comment by nzakas** [comment]
@TobiasBales rule changes require a team member champion its inclusion (per our [guidelines](http://eslint.org/docs/developer-guide/contributing/rule-changes#accepting-a-rule-change)).

**Comment by hudson155** [comment]
With Angular 2 being officially released, this would be extremely useful

**Comment by fczuardi** [comment]
An example of what @aubergine10 [described above](https://github.com/eslint/eslint/issues/6505#issuecomment-228630569) can be seen in the [max-len rule](http://eslint.org/docs/rules/max-len), that supports exceptions with ignorePattern, ignoreUrls, etc…

**Comment by rhbecker** [comment]
The `argsIgnorePattern` option for `no-unused-vars` is very similar to this, and similarly useful to those writing express apps: `"no-unused-vars": ["error", { "argsIgnorePattern": "next" }]`.

Elaborating on the use cases for adding a similar option to `no-param-reassign`, Express docs suggest [setting per-request values within middleware](http://expressjs.com/en/api.html#res.locals) by modifying `res.locals`.

**Comment by jasonmorita** [comment]
You can also use lodash `assignIn` like:

```
assignIn(req, { someNewObj });
```

Still mutates, but no comment needed.

https://lodash.com/docs/4.17.2#assignIn

**Comment by hudson155** [comment]
@jasonmorita I'd rather have my linter pick up on this than use it as workaround for something my linter doesn't support.

**Comment by platinumazure** [comment]
Okay, this needs a champion and I'll step up.

**Comment by ljharb** [comment]
Just stumbled across this; the airbnb config [will be enabling this option by default](https://github.com/airbnb/javascript/issues/1217) as soon as it's released, to cover express and angular apps in particular.

**Comment by TobiasBales** [comment]
@platinumazure that is great. You have seen the PR I created that implemented the rule addition and included tests, right?

@ljharb glad to hear :D

**Comment by vitorbal** [comment]
Is there a final proposal on how the schema for this option would look like?

**Comment by platinumazure** [comment]
@vitorbal I'll throw this out:

```json
"rules": {
    "no-param-reassign": [
        "error",
        { ignorePropertyAssignmentsFor: ["foo", "bar"] }
    ]
}
```

This could later be complemented with an `ignoreAssignmentsFor` (for per-identifier full reassignments), if desired. I also propose that the new option would not be usable with `properties: true`.

Thoughts?

**Comment by platinumazure** [comment]
Adding "tsc agenda" as a TSC member has expressed opposition.

**TSC Summary:** This proposal is for a new rule option for `no-param-reassign` (name still being bikeshedded). In Angular, Express, and Chrome extensions (among other places), it is common and expected to mutate properties of function arguments. The new rule option takes in a list of identifiers (strings); when any of those identifiers shows up as a function parameter, assigning to a property of that identifier will not be flagged even with `properties: true`. Users can work around this by either fully disabling the rule or by using `eslint-disable` comments. Problem is that users may need to use lots of disable comments in order to realize the benefits of the rule in other situations.

@ilyavolodin (TSC member) has already objected to the proposal.

**TSC Question:** Should we accept this rule?

**Comment by rhbecker** [comment]
@platinumazure 

Regarding your proposed implementation, I had actually hoped to maintain `{ "props": true }` while providing a means for ignoring particular properties for particular identifiers.

As I stated earlier, my use case is related to the Express docs suggestion for [setting per-request values within middleware](http://expressjs.com/en/api.html#res.locals) by modifying `res.locals`. Ideally, my linter would allow that particular property modification while still barking at me for modifying other properties of `res`.

My other comment on your proposal is that patterns might prove more flexible and convenient than lists. I previously alluded to the `argsIgnorePattern` option for `no-unused-vars` as a potential model in that regard. Presumably, if pattern support was already a good enough design to implement for one rule, it would be worth considering here.

Perhaps these factors introduce undesirable complexities, in terms of implementation? I'm only commenting on what would be desirable from my perspective.

Finally, what was the nature of the already raised objection?

**Comment by rhbecker** [comment]
With respect to the two issues I raised in the previous comment, is something like this too complicated?

```javascript
"rules": {
    "no-param-reassign": [
        "error",
        {
            "paramsIgnorePattern": {
                "^_": ".+", // ignore all properties for params starting with underscore
                "^res$": "^local$" // ignore only the "local" property of the res param
            }
        }
    ]
}
```

**Comment by ljharb** [comment]
@rhbecker allowing regular expressions (as opposed to simply wildcards) in user-supplied config imo is a very bad idea; it's hard to sanitize, and an improperly constructed regex could crash the process.

**Comment by not-an-aardvark** [comment]
@ljharb While I'm not sure I agree with adding a `paramsIgnorePatterns` option that allows regular expressions, I don't understand your concerns with regular expressions in general. There are a lot of rules that currently accept regular expression options. It can be sanitized by just using `new RegExp(userConfig)`.

**Comment by ljharb** [comment]
@not-an-aardvark we can discuss it on a separate medium; but what rules currently allow more than "*" wildcards?

**Comment by not-an-aardvark** [comment]
[`capitalized-comments`](http://eslint.org/docs/rules/capitalized-comments), [`default-case`](http://eslint.org/docs/rules/default-case), [`dot-notation`](http://eslint.org/docs/rules/dot-notation), [`handle-callback-err`](http://eslint.org/docs/rules/handle-callback-err), [`id-match`](http://eslint.org/docs/rules/id-match), [`line-comment-position`](http://eslint.org/docs/rules/line-comment-position), [`no-fallthrough`](http://eslint.org/docs/rules/no-fallthrough), and [`valid-jsdoc`](http://eslint.org/docs/rules/valid-jsdoc).

**Comment by hudson155** [comment]
I would love regular expressions, but my concern is that someone who wants to match "foo" has to put `"^foo$"` rather than just `"foo"`.

**Comment by rhbecker** [comment]
@ljharb @not-an-aardvark 

Why is no one noticing that I keep referring to an example where this already exists?

> `argsIgnorePattern` option for `no-unused-vars`

That same rule also has `caughtErrorsIgnorePattern` and `varsIgnorePattern` options, which accept regex values.

I can survive without regex if it's truly counter to where you'd want to go with eslint, but I felt it was likely safe to suggest copying a design approach already implemented elsewhere.

**Comment by rhbecker** [comment]
@hudson155 I agree it's slightly less convenient to write as a regex, but I'd like to think that anyone who cares enough to bother specifying an exception to a linting rule that could otherwise be shut off might be willing to go the extra "mile". Also, while your example happens to be less convenient, certainly there'd be other examples where a regex would be more convenient than having to list out several "literal" exceptions.

**Comment by rhbecker** [comment]
@not-an-aardvark I figured what I proposed might not be idea ... was mostly trying to exhibit what I was describing, in terms of capabilities. I bet others here are more well equipped to design a friendlier schema. If what I'm after proves too edge-case-y, I do think the originally proposed schema is better than nothing.

**Comment by rhbecker** [comment]
@platinumazure Would you mind clarifying what you meant by your statement ...

> I also propose that the new option would not be usable with `properties: true`.

1. Is it safe to assume that you're actually referring to [`props` rather than `properties`](http://eslint.org/docs/rules/no-param-reassign#props)?
2. By "new option" are you referring to your proposed `ignorePropertyAssignmentsFor` option, or your proposed `ignoreAssignmentsFor` option?
3. Whichever option you were referring to, why would it conflict with `props` being set to `true`?
4. Am I right to assume that with the example presented in your proposed schema (which leverages your proposed `ignorePropertyAssignmentsFor` option), eslint would no longer error on reassignment of `foo.x` or `bar.x`, but _would_ error on `baz.x`? Wouldn't I need `props` set to `true` for this new option to even come into play?

**Comment by mysticatea** [comment]
I remembered about [new-cap](http://eslint.org/docs/rules/new-cap) rule.
The rule has the simple regexp option to except property accesses.

**Comment by rhbecker** [comment]
@platinumazure: thanks for the clarifications!

**Comment by christianbundy** [comment]
Just to document, the solution for the issue example (and for anyone else dealing with Angular and `$scope`):

```js
/*eslint no-param-reassign: ["error", { "props": true, "ignorePropertyModificationsFor": ["$scope"] }]*/

app.controller('MyCtrl', function($scope) {
  $scope.something = true;
});
```

**Comment by peter-mouland** [comment]
sorry for commenting on a closed issue - but i just wanted to make sure i'm _doing it right_ before going further.

I have `eslint@3.18.0`
the rule is: `    "no-param-reassign": ["error", { "ctx": false }],`
I would like to `don't error when altering ctx`
I get the error: `Value "[object Object]" no (or more than one) schemas match.`

Have I done something stupid?

**Comment by christianbundy** [comment]
@peter-mouland Can you double-check your syntax for me? It looks like you're doing this:

```json
{
  "no-param-reassign": ["error", {
    "ctx": false
  }]
}
```

Try this instead:

```js
{
  "no-param-reassign": ["error", {
    "props": true,
    "ignorePropertyModificationsFor": ["ctx"]
  }]
}
```
